vhd/lib/libvhd.c

```diff
@@ -119,6 +120,128 @@ old_clear_bit(volatile char *addr, int nr)
 	((uint32_t *)addr)[nr >> 5] &= ~(1 << (nr & 31));
 }
 
+static int
+makedev_from_file(const char *file, dev_t *dev)
+{
+	*dev = 0;
+
+	FILE *f = fopen(file, "r");
+	if (!f)
+		return -1;
+
+	unsigned int dev_major;
+	unsigned int dev_minor;
+	const int ret = fscanf(f, "%u:%u", &dev_major, &dev_minor);
+	fclose(f);
+
+	if (ret != 2)
+		return -1;
+
+	*dev = makedev(dev_major, dev_minor);
+	return 0;
+}
+
+static int
+is_drbd_device(const struct stat *stats)
+{
+	static const int drbd_major = 147;
+
+	if (!S_ISBLK(stats->st_mode))
+		return 0;
+
+	const unsigned int dev_major = major(stats->st_rdev);
+	if (dev_major != drbd_major)
+		return 0;
+
+	return 1;
+}
+
+static int
+drbd_is_up_to_date(const struct stat *stats)
+{
+	static const char up_to_date_prefix[] = "UpToDate";
+
+	if (!is_drbd_device(stats))
+		return 0;
+
+	char cmd[64];
+	snprintf(cmd, sizeof cmd, "drbdsetup dstate %u 2> /dev/null", minor(stats->st_rdev));
+
+	int ok = 0;
+
+	FILE *stream = popen(cmd, "r");
+	char line[1024] = {0};
+	if (
+		stream && fscanf(stream, "%1023[^\n]", line) == 1 &&
+		!strncmp(line, up_to_date_prefix, sizeof up_to_date_prefix - 1)
+	)
+		ok = 1;
+
+	if (stream)
+		pclose(stream);
+
+	return ok;
+}
+
+static int
+drbd_to_mapper(const struct stat *stats, dev_t *dev)
+{
+	*dev = 0;
+
+	if (!is_drbd_device(stats))
+		return -1;
+
+	/* Ok we can try to search a valid slave device. */
+	/* Note: If it's a diskless DRBD device, there is no slave. */
+	const unsigned int dev_major = major(stats->st_rdev);
+	const unsigned int dev_minor = minor(stats->st_rdev);
+	char slaves_dir[PATH_MAX];
+	snprintf(slaves_dir, sizeof slaves_dir, "/sys/dev/block/%u:%u/slaves", dev_major, dev_minor);
+
+	DIR *dir = opendir(slaves_dir);
+	if (!dir)
+		return -1;
+
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		/* Find a valid symlink to a slave. */
+		#ifdef _DIRENT_HAVE_D_TYPE
+			if (ent->d_type == DT_LNK) {
+				/* Ok here, we have a symlink! */
+			} else if (ent->d_type != DT_UNKNOWN)
+				continue;
+			else
+		#endif
+		{
+			 struct stat slave_stats;
+			 stat(ent->d_name, &slave_stats);
```

Comment:
You don't check the return value of stat here, this fails a Coverity check as the stat struct may not have been populated.

"check_return: Calling stat(ent->d_name, &slave_stats) without checking return value. This library function may fail and return an error code"

Reply:
Oops, yes that's right. I just fixed it!